jstz_api/src/urlpattern/mod.rs

```diff
@@ -0,0 +1,576 @@
+            Ok(Some(input)) => Ok(input),
+            Ok(None) => Err(JsError::from_native(JsNativeError::error())),
+        }
+    }
+
+    // We do not support options (ignoreCase), as it is not supported in Deno
+    // nor in `urlpattern` crate. There is an open PR for supporting it in
+    // `urlpattern`, and we could use it when it gets merged.
+    pub fn new(
+        _this: &JsNativeObject<Self>,
+        input: UrlPatternInput,
+        base_url: Option<String>,
+        _context: &mut Context<'_>,
+    ) -> JsResult<Self> {
+        let UrlPatternInput(stringorinit) = input;
+        let urlpatterninit = urlpattern::quirks::process_construct_pattern_input(
+            stringorinit,
+            base_url.as_deref(),
+        )
+        .map_err(|_| {
+            JsError::from_native(
+                JsNativeError::typ().with_message("Failed to build UrlPatternInit"),
+            )
+        })?;
+        let url_pattern = InnerUrlPattern::parse(urlpatterninit).map_err(|_| {
+            JsError::from_native(
+                JsNativeError::typ().with_message("Failed to parse UrlPatternInit"),
+            )
+        })?;
+
+        Ok(Self { url_pattern })
+    }
+
+    pub fn protocol(&self) -> String {
+        String::from(self.url_pattern.protocol())
+    }
+
+    pub fn username(&self) -> String {
+        String::from(self.url_pattern.username())
+    }
+
+    pub fn password(&self) -> String {
+        String::from(self.url_pattern.password())
+    }
+
+    pub fn hostname(&self) -> String {
+        String::from(self.url_pattern.hostname())
+    }
+
+    pub fn port(&self) -> String {
+        String::from(self.url_pattern.port())
+    }
+
+    pub fn pathname(&self) -> String {
+        String::from(self.url_pattern.pathname())
+    }
+
+    pub fn search(&self) -> String {
+        String::from(self.url_pattern.search())
+    }
+
+    pub fn hash(&self) -> String {
+        String::from(self.url_pattern.hash())
+    }
+
+    pub fn test(
+        &self,
+        input: UrlPatternInput,
+        base_url: Option<String>,
+    ) -> JsResult<bool> {
+        let UrlPatternInput(string_or_init) = input;
+        let (url_pattern_match_input, _) = Self::process_input(string_or_init, base_url)?;
+
+        self.url_pattern
+            .test(url_pattern_match_input)
+            .map_err(|e| JsNativeError::typ().with_message(e.to_string()).into())
+    }
+
+    pub fn exec(
+        &self,
+        input: UrlPatternInput,
+        base_url: Option<String>,
+    ) -> JsResult<Option<UrlPatternResult>> {
+        let UrlPatternInput(string_or_init) = input;
+        let (url_pattern_match_input, (string_or_init, base_url)) =
+            Self::process_input(string_or_init, base_url)?;
+
+        let mut inputs: Vec<UrlPatternInput> = Vec::new();
+        inputs.push(UrlPatternInput(string_or_init));
+        if let Some(base_url) = base_url {
+            inputs.push(UrlPatternInput(InnerStringOrQuirksInit::String(base_url)));
```

Comment:
Would be best to implement `From` trait for the two enums in `UrlPatternInput`.

Reply:
Implemented, and used it here.